src/Ocelot/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinder.cs

```diff
@@ -1,161 +1,132 @@
+        template = EscapeExceptBraces(template);
+        string regexPattern = $"^{regex.Replace(template, match => $"(?<{match.Groups[1].Value}>[^&]*)")}";
+        
+        string testedPath = ShouldSkipQuery(query, template) ? path : $"{path}{query}";
+        Match match = Regex.Match(testedPath, regexPattern);
+        List<Group> foundGroups = match.Groups.Cast<Group>().Skip(1).ToList();
+        
+        if (foundGroups.Count > 0)
         {
-            return path.Length == 1 || path.Length == 0;
+            return foundGroups;
         }
 
-        private static string GetPlaceholderValue(string urlPathTemplate, string query, string variableName, string urlPath, int counterForUrl, char delimiter)
+        if (!IsCatchAllPath(template))
         {
-            if (counterForUrl >= urlPath.Length)
-            {
-                return string.Empty;
-            }
-
-            if ( urlPath[counterForUrl] == '/')
-            {
-                counterForUrl++;
-            }
-
-            var positionOfNextSlash = urlPath.IndexOf(delimiter, counterForUrl);
-
-            if (positionOfNextSlash == -1 || (urlPathTemplate.Trim(delimiter).EndsWith(variableName) && string.IsNullOrEmpty(query)))
-            {
-                positionOfNextSlash = urlPath.Length;
-            }
-
-            var variableValue = urlPath.Substring(counterForUrl, positionOfNextSlash - counterForUrl);
-
-            return variableValue;
+            return foundGroups;
         }
 
-        private static string GetPlaceholderName(string urlPathTemplate, int counterForTemplate)
-        {
-            var postitionOfPlaceHolderClosingBracket = urlPathTemplate.IndexOf('}', counterForTemplate) + 1;
-
-            var variableName = urlPathTemplate.Substring(counterForTemplate, postitionOfPlaceHolderClosingBracket - counterForTemplate);
-
-            return variableName;
-        }
+        // Append a trailing slash to the path if it is a catch-all path
+        match = Regex.Match($"{testedPath}/", regexPattern);
+        return match.Groups.Cast<Group>().Skip(1).ToList();
+    }
 
-        private static int GetNextCounterPosition(string urlTemplate, int counterForTemplate, char delimiter)
-        {
-            var closingPlaceHolderPositionOnTemplate = urlTemplate.IndexOf(delimiter, counterForTemplate);
-            return closingPlaceHolderPositionOnTemplate + 1;
-        }
+    /// <summary>
+    /// Checks if the path template contains a catch-all query parameter.
+    /// It means that the path template ends with a question mark and a placeholder.
+    /// And no other placeholders are present in the path template.
+    /// </summary>
+    /// <param name="template">The path template.</param>
+    /// <returns>True if it matches and the found placeholder.</returns>
+    private static (bool IsMatch, string Placeholder) IsCatchAllQuery(string template)
```

Comment:
Styling and convention issue of using `Tuple` in a method which should simply return `bool` only.
```suggestion
    private static bool IsCatchAllQuery(string template, out string placeholder)
```

Reply:
Question of taste again